sdk/iptables/iptables.go

```diff
@@ -0,0 +1,158 @@
+type Config struct {
+	// ProxyUserID is the user ID of the proxy process.
+	ProxyUserID string
+
+	// ProxyInboundPort is the port of the proxy's inbound listener.
+	ProxyInboundPort int
+
+	// ProxyInboundPort is the port of the proxy's outbound listener.
+	ProxyOutboundPort int
+
+	// IptablesProvider is the Provider that will apply iptables rules.
+	IptablesProvider Provider
+}
+
+// Provider is an interface for executing iptables rules.
+type Provider interface {
+	// AddRule adds a rule without executing it.
+	AddRule(name string, args ...string)
+	// ApplyRules executes rules that have been added via AddRule.
+	ApplyRules() error
+	// Rules returns the list of rules that have been added but not applied yet.
+	Rules() []string
+}
+
+// iptablesExecutor implements IptablesProvider using exec.Cmd.
+type iptablesExecutor struct {
+	commands []*exec.Cmd
+}
+
+func (i *iptablesExecutor) AddRule(name string, args ...string) {
+	i.commands = append(i.commands, exec.Command(name, args...))
+}
+
+func (i *iptablesExecutor) ApplyRules() error {
+	_, err := exec.LookPath("iptables")
+	if err != nil {
+		return err
+	}
+
+	for _, cmd := range i.commands {
+		err := cmd.Run()
+		if err != nil {
+			output, err := cmd.CombinedOutput()
+
+			if err != nil {
+				return fmt.Errorf("failed to run command: %s, err: %v", cmd.String(), err)
+			}
+
+			return fmt.Errorf("failed to run command: %s, err: %v, output: %s", cmd.String(), err, string(output))
+		}
+	}
+
+	return nil
+}
+
+func (i *iptablesExecutor) Rules() []string {
+	var rules []string
+	for _, cmd := range i.commands {
+		rules = append(rules, cmd.String())
+	}
+
+	return rules
+}
+
+// Setup will set up iptables interception and redirection rules
+// based on the configuration provided in cfg.
+// This implementation was inspired by
+// https://github.com/openservicemesh/osm/blob/650a1a1dcf081ae90825f3b5dba6f30a0e532725/pkg/injector/iptables.go
+func Setup(cfg Config) error {
+	if cfg.IptablesProvider == nil {
+		cfg.IptablesProvider = &iptablesExecutor{}
+	}
+
+	err := validateConfig(cfg)
+	if err != nil {
+		return err
+	}
+
+	// Create chains we will use for redirection.
+	chains := []string{ProxyInboundChain, ProxyInboundRedirectChain, ProxyOutputChain, ProxyOutputRedirectChain}
+	for _, chain := range chains {
+		cfg.IptablesProvider.AddRule("iptables", "-t", "nat", "-N", chain)
+	}
+
+	// Configure outbound rules.
+	{
+		// Redirects outbound TCP traffic hitting PROXY_REDIRECT chain to Envoy's outbound listener port.
+		cfg.IptablesProvider.AddRule("iptables", "-t", "nat", "-A", ProxyOutputRedirectChain, "-p", "tcp", "-j", "REDIRECT", "--to-port", strconv.Itoa(cfg.ProxyOutboundPort))
+
+		// For outbound TCP traffic jump from OUTPUT chain to PROXY_OUTPUT chain.
+		cfg.IptablesProvider.AddRule("iptables", "-t", "nat", "-A", "OUTPUT", "-p", "tcp", "-j", ProxyOutputChain)
+
+		// Don't redirect Envoy traffic back to itself, return it to the next chain for processing.
```

Comment:
```suggestion
		// Don't redirect proxy traffic back to itself, return it to the next chain for processing.
```

This library doesn't need to be Envoy specific, so we can use a more generic word here. Same applies to the other Envoy references in this file.